File: sdk/python/aip7h3/protocol.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
from dataclasses import dataclass
from typing import Any, Dict, Optional, TypedDict

try:
    from cryptography.hazmat.primitives import serialization
    from cryptography.hazmat.primitives.asymmetric.ed25519 import (
        Ed25519PrivateKey,
        Ed25519PublicKey,
    )

    _HAS_CRYPTOGRAPHY = True
except Exception:  # pragma: no cover - optional dependency
    _HAS_CRYPTOGRAPHY = False


class ProtocolHeader(TypedDict, total=False):
    version: str
    messageId: str
    timestampMs: int
    ttlMs: int
    sender: str
    recipient: str
    nonce: str


class ProtocolBody(TypedDict, total=False):
    intent: str
    content: str
    capability: str
    correlationId: str


class ProtocolSignature(TypedDict):
    alg: str
    keyId: str
    value: str


class ProtocolEnvelope(TypedDict, total=False):
    header: ProtocolHeader
    body: ProtocolBody
    signature: ProtocolSignature


@dataclass(frozen=True)
class ProtocolDiagnostic:
    level: str
    message: str


def _json_string(value: Any) -> str:
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def _base64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def _base64url_decode(value: str) -> bytes:
    padding = "=" * ((4 - (len(value) % 4)) % 4)
    return base64.urlsafe_b64decode(value + padding)


def canonicalize_envelope(envelope: Dict[str, Any]) -> str:
    header = envelope["header"]
    body = envelope["body"]

    body_parts = []
    if "capability" in body and body["capability"] is not None:
        body_parts.append(f'"capability":{_json_string(body["capability"])}')
    body_parts.append(f'"content":{_json_string(body["content"])}')
    if "correlationId" in body and body["correlationId"] is not None:
        body_parts.append(f'"correlationId":{_json_string(body["correlationId"])}')
    body_parts.append(f'"intent":{_json_string(body["intent"])}')

    header_parts = [
        f'"messageId":{_json_string(header["messageId"])}',
        f'"nonce":{_json_string(header["nonce"])}',
    ]
    if "recipient" in header and header["recipient"] is not None:
        header_parts.append(f'"recipient":{_json_string(header["recipient"])}')
    header_parts.extend(
        [
            f'"sender":{_json_string(header["sender"])}',
            f'"timestampMs":{int(header["timestampMs"])}',
            f'"ttlMs":{int(header["ttlMs"])}',
            f'"version":{_json_string(header["version"])}',
        ]
    )

    return (
        f'{{"body":{{{",".join(body_parts)}}},"header":{{{",".join(header_parts)}}}}}'
    )


def sign_canonical_payload_hmac(canonical_payload: str, secret: str) -> str:
    digest = hmac.new(
        secret.encode("utf-8"), canonical_payload.encode("utf-8"), hashlib.sha256
    ).digest()
    return _base64url(digest)


def verify_canonical_payload_hmac(
    canonical_payload: str, signature: str, secret: str
) -> bool:
    expected = sign_canonical_payload_hmac(canonical_payload, secret)
    return hmac.compare_digest(signature, expected)


def sign_canonical_payload_ed25519(
    canonical_payload: str, private_key_pkcs8_base64url: str
) -> str:
    if not _HAS_CRYPTOGRAPHY:
        raise RuntimeError("cryptography package is required for ED25519 operations")

    private_der = _base64url_decode(private_key_pkcs8_base64url)
    private_key = serialization.load_der_private_key(private_der, password=None)
    if not isinstance(private_key, Ed25519PrivateKey):
        raise ValueError("Invalid ED25519 private key")
    signature = private_key.sign(canonical_payload.encode("utf-8"))
    return _base64url(signature)


def verify_canonical_payload_ed25519(
    canonical_payload: str, signature: str, public_key_spki_base64url: str
) -> bool:
    if not _HAS_CRYPTOGRAPHY:
        raise RuntimeError("cryptography package is required for ED25519 operations")

    public_der = _base64url_decode(public_key_spki_base64url)
    public_key = serialization.load_der_public_key(public_der)
    if not isinstance(public_key, Ed25519PublicKey):
        return False

    try:
        public_key.verify(
            _base64url_decode(signature), canonical_payload.encode("utf-8")
        )
        return True
    except Exception:
        return False


def sign_envelope_hmac(
    envelope: Dict[str, Any], secret: str, key_id: str = "python-dev-key"
) -> Dict[str, Any]:
    canonical = canonicalize_envelope(envelope)
    signature = sign_canonical_payload_hmac(canonical, secret)
    return {
        "header": dict(envelope["header"]),
        "body": dict(envelope["body"]),
        "signature": {"alg": "HS256", "keyId": key_id, "value": signature},
    }


def verify_envelope_hmac(envelope: Dict[str, Any], secret: str) -> bool:
    signature = envelope.get("signature")
    if not signature:
        return False
    if signature.get("alg") != "HS256":
        return False
    unsigned = {"header": envelope["header"], "body": envelope["body"]}
    canonical = canonicalize_envelope(unsigned)
    return verify_canonical_payload_hmac(canonical, signature.get("value", ""), secret)


def sign_envelope_ed25519(
    envelope: Dict[str, Any],
    private_key_pkcs8_base64url: str,
    key_id: str = "python-ed25519-key",
) -> Dict[str, Any]:
    canonical = canonicalize_envelope(envelope)
    signature = sign_canonical_payload_ed25519(canonical, private_key_pkcs8_base64url)
    return {
        "header": dict(envelope["header"]),
        "body": dict(envelope["body"]),
        "signature": {"alg": "ED25519", "keyId": key_id, "value": signature},
    }


def verify_envelope_ed25519(
    envelope: Dict[str, Any], public_key_spki_base64url: str
) -> bool:
    signature = envelope.get("signature")
    if not signature:
        return False
    if signature.get("alg") != "ED25519":
        return False
    unsigned = {"header": envelope["header"], "body": envelope["body"]}
    canonical = canonicalize_envelope(unsigned)
    return verify_canonical_payload_ed25519(
        canonical, signature.get("value", ""), public_key_spki_base64url
    )


def encode_envelope_compact(envelope: Dict[str, Any]) -> str:
    header = envelope["header"]
    body = envelope["body"]
    compact: Dict[str, Any] = {
        "v": header["version"],
        "mid": header["messageId"],
        "ts": header["timestampMs"],
        "ttl": header["ttlMs"],
        "s": header["sender"],
        "n": header["nonce"],
        "i": body["intent"],
        "c": body["content"],
    }
    if "recipient" in header and header["recipient"] is not None:
        compact["r"] = header["recipient"]
    if "capability" in body and body["capability"] is not None:
        compact["cap"] = body["capability"]
    if "correlationId" in body and body["correlationId"] is not None:
        compact["cid"] = body["correlationId"]

    signature = envelope.get("signature")
    if signature:
        compact["sig"] = {
            "a": signature.get("alg", "HS256"),
            "k": signature["keyId"],
            "v": signature["value"],
        }

    return _json_string(compact)


def decode_envelope(raw: str) -> Dict[str, Any]:
    parsed = json.loads(raw)
    if "header" in parsed and "body" in parsed:
        return parsed

    if parsed.get("v") != "aip/0.1" or "mid" not in parsed or "i" not in parsed:
        raise ValueError("Envelope JSON shape is not recognized")

    header: Dict[str, Any] = {
        "version": parsed["v"],
        "messageId": parsed["mid"],
        "timestampMs": parsed["ts"],
        "ttlMs": parsed["ttl"],
        "sender": parsed["s"],
        "nonce": parsed["n"],
    }
    if "r" in parsed:
        header["recipient"] = parsed["r"]

    body: Dict[str, Any] = {
        "intent": parsed["i"],
        "content": parsed["c"],
    }
    if "cap" in parsed:
        body["capability"] = parsed["cap"]
    if "cid" in parsed:
        body["correlationId"] = parsed["cid"]

    envelope: Dict[str, Any] = {
        "header": header,
        "body": body,
    }
    if parsed.get("sig"):
        envelope["signature"] = {
            "alg": parsed["sig"].get("a", "HS256"),
            "keyId": parsed["sig"]["k"],
            "value": parsed["sig"]["v"],
        }
    return envelope


def validate_envelope(
    envelope: Dict[str, Any], now_ms: Optional[int] = None
) -> list[ProtocolDiagnostic]:
    header = envelope["header"]
    body = envelope["body"]
    current = int(now_ms if now_ms is not None else 0)
    diagnostics: list[ProtocolDiagnostic] = []

    if header.get("version") != "aip/0.1":
        diagnostics.append(
            ProtocolDiagnostic(
                level="error",
                message=f"Unsupported protocol version '{header.get('version')}'",
            )
        )
    if not str(header.get("messageId", "")).strip():
        diagnostics.append(
            ProtocolDiagnostic(level="error", message="Missing messageId")
        )
    if not str(header.get("sender", "")).strip():
        diagnostics.append(
            ProtocolDiagnostic(level="error", message="Missing sender identity")
        )
    if int(header.get("ttlMs", 0)) <= 0:
        diagnostics.append(
            ProtocolDiagnostic(level="error", message="ttlMs must be greater than zero")
        )

    if now_ms is not None:
        if int(header.get("timestampMs", 0)) + int(header.get("ttlMs", 0)) < current:
            diagnostics.append(
                ProtocolDiagnostic(level="error", message="Message TTL expired")
            )

    if not str(body.get("content", "")).strip():
        diagnostics.append(
            ProtocolDiagnostic(level="warning", message="Empty content payload")
        )

    return diagnostics

File: src/protocol.js
const textEncoder = new TextEncoder();
const HMAC_KEY_CACHE_LIMIT = 256;
const hmacKeyCache = new Map();
const ED25519_KEY_CACHE_LIMIT = 256;
const ed25519PrivateKeyCache = new Map();
const ed25519PublicKeyCache = new Map();
function getBufferLike() {
    const candidate = globalThis;
    return candidate.Buffer ?? null;
}
function serializeHeaderCanonical(header) {
    const parts = [
        `"messageId":${JSON.stringify(header.messageId)}`,
        `"nonce":${JSON.stringify(header.nonce)}`,
    ];
    if (header.recipient !== undefined) {
        parts.push(`"recipient":${JSON.stringify(header.recipient)}`);
    }
    parts.push(`"sender":${JSON.stringify(header.sender)}`);
    parts.push(`"timestampMs":${header.timestampMs}`);
    parts.push(`"ttlMs":${header.ttlMs}`);
    parts.push(`"version":${JSON.stringify(header.version)}`);
    return `{${parts.join(',')}}`;
}
function serializeBodyCanonical(body) {
    const parts = [];
    if (body.capability !== undefined) {
        parts.push(`"capability":${JSON.stringify(body.capability)}`);
    }
    parts.push(`"content":${JSON.stringify(body.content)}`);
    if (body.correlationId !== undefined) {
        parts.push(`"correlationId":${JSON.stringify(body.correlationId)}`);
    }
    parts.push(`"intent":${JSON.stringify(body.intent)}`);
    return `{${parts.join(',')}}`;
}
function getCachedHmacKey(secret) {
    const cacheKey = `hs256:${secret}`;
    const cached = hmacKeyCache.get(cacheKey);
    if (cached)
        return cached;
    if (hmacKeyCache.size >= HMAC_KEY_CACHE_LIMIT) {
        hmacKeyCache.clear();
    }
    const subtle = requireCryptoSubtle();
    const imported = subtle
        .importKey('raw', textEncoder.encode(secret), { name: 'HMAC', hash: 'SHA-256' }, false, ['sign', 'verify'])
        .catch((error) => {
        hmacKeyCache.delete(cacheKey);
        throw error;
    });
    hmacKeyCache.set(cacheKey, imported);
    return imported;
}
function getCachedEd25519PrivateKey(privateKeyPkcs8Base64Url) {
    const cacheKey = `ed25519:pkcs8:${privateKeyPkcs8Base64Url}`;
    const cached = ed25519PrivateKeyCache.get(cacheKey);
    if (cached)
        return cached;
    if (ed25519PrivateKeyCache.size >= ED25519_KEY_CACHE_LIMIT) {
        ed25519PrivateKeyCache.clear();
    }
    const subtle = requireCryptoSubtle();
    const imported = subtle
        .importKey('pkcs8', toArrayBuffer(fromBase64Url(privateKeyPkcs8Base64Url)), { name: 'Ed25519' }, false, ['sign'])
        .catch((error) => {
        ed25519PrivateKeyCache.delete(cacheKey);
        throw error;
    });
    ed25519PrivateKeyCache.set(cacheKey, imported);
    return imported;
}
function getCachedEd25519PublicKey(publicKeySpkiBase64Url) {
    const cacheKey = `ed25519:spki:${publicKeySpkiBase64Url}`;
    const cached = ed25519PublicKeyCache.get(cacheKey);
    if (cached)
        return cached;
    if (ed25519PublicKeyCache.size >= ED25519_KEY_CACHE_LIMIT) {
        ed25519PublicKeyCache.clear();
    }
    const subtle = requireCryptoSubtle();
    const imported = subtle
        .importKey('spki', toArrayBuffer(fromBase64Url(publicKeySpkiBase64Url)), { name: 'Ed25519' }, false, ['verify'])
        .catch((error) => {
        ed25519PublicKeyCache.delete(cacheKey);
        throw error;
    });
    ed25519PublicKeyCache.set(cacheKey, imported);
    return imported;
}
function toBase64Url(bytes) {
    const bufferLike = getBufferLike();
    const base64 = bufferLike ? bufferLike.from(bytes).toString('base64') : btoa(String.fromCharCode(...bytes));
    return base64
        .replace(/\+/g, '-')
        .replace(/\//g, '_')
        .replace(/=+$/g, '');
}
function fromBase64Url(value) {
    const padded = value
        .replace(/-/g, '+')
        .replace(/_/g, '/')
        .padEnd(Math.ceil(value.length / 4) * 4, '=');
    const bufferLike = getBufferLike();
    if (bufferLike) {
        return new Uint8Array(bufferLike.from(padded, 'base64'));
    }
    const binary = atob(padded);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i += 1) {
        bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
}
function toArrayBuffer(bytes) {
    return bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength);
}
export function canonicalizeEnvelope(envelope) {
    return `{"body":${serializeBodyCanonical(envelope.body)},"header":${serializeHeaderCanonical(envelope.header)}}`;
}
function requireCryptoSubtle() {
    if (typeof crypto === 'undefined' || !crypto.subtle) {
        throw new Error('Web Crypto API is not available in this runtime');
    }
    return crypto.subtle;
}
async function hmacSign(payload, secret) {
    const subtle = requireCryptoSubtle();
    const key = await getCachedHmacKey(secret);
    const signature = await subtle.sign('HMAC', key, textEncoder.encode(payload));
    return toBase64Url(new Uint8Array(signature));
}
async function hmacVerify(payload, signature, secret) {
    const subtle = requireCryptoSubtle();
    const key = await getCachedHmacKey(secret);
    const signatureBytes = fromBase64Url(signature);
    return subtle.verify('HMAC', key, signatureBytes.buffer, textEncoder.encode(payload));
}
async function ed25519Sign(payload, privateKeyPkcs8Base64Url) {
    const subtle = requireCryptoSubtle();
    const key = await getCachedEd25519PrivateKey(privateKeyPkcs8Base64Url);
    const signature = await subtle.sign('Ed25519', key, textEncoder.encode(payload));
    return toBase64Url(new Uint8Array(signature));
}
async function ed25519Verify(payload, signature, publicKeySpkiBase64Url) {
    const subtle = requireCryptoSubtle();
    const key = await getCachedEd25519PublicKey(publicKeySpkiBase64Url);
    const signatureBytes = fromBase64Url(signature);
    return subtle.verify('Ed25519', key, signatureBytes.buffer, textEncoder.encode(payload));
}
export async function signCanonicalPayloadHmac(payload, secret) {
    return hmacSign(payload, secret);
}
export async function verifyCanonicalPayloadHmac(payload, signature, secret) {
    return hmacVerify(payload, signature, secret);
}
export async function generateEd25519KeypairBase64Url() {
    const subtle = requireCryptoSubtle();
    const pair = await subtle.generateKey({ name: 'Ed25519' }, true, ['sign', 'verify']);
    const privateKeyRaw = await subtle.exportKey('pkcs8', pair.privateKey);
    const publicKeyRaw = await subtle.exportKey('spki', pair.publicKey);
    return {
        privateKey: toBase64Url(new Uint8Array(privateKeyRaw)),
        publicKey: toBase64Url(new Uint8Array(publicKeyRaw)),
    };
}
export async function signCanonicalPayloadEd25519(payload, privateKeyPkcs8Base64Url) {
    return ed25519Sign(payload, privateKeyPkcs8Base64Url);
}
export async function verifyCanonicalPayloadEd25519(payload, signature, publicKeySpkiBase64Url) {
    return ed25519Verify(payload, signature, publicKeySpkiBase64Url);
}
export async function signEnvelopeHmac(envelope, secret, keyId = 'local-dev-key') {
    const payload = canonicalizeEnvelope(envelope);
    const signature = await hmacSign(payload, secret);
    return {
        ...envelope,
        signature: {
            alg: 'HS256',
            keyId,
            value: signature,
        },
    };
}
export async function verifyEnvelopeHmac(envelope, secret) {
    if (!envelope.signature)
        return false;
    if (envelope.signature.alg !== 'HS256')
        return false;
    const unsigned = {
        header: envelope.header,
        body: envelope.body,
    };
    const payload = canonicalizeEnvelope(unsigned);
    return hmacVerify(payload, envelope.signature.value, secret);
}
export async function signEnvelopeEd25519(envelope, privateKeyPkcs8Base64Url, keyId = 'local-ed25519-key') {
    const payload = canonicalizeEnvelope(envelope);
    const signature = await ed25519Sign(payload, privateKeyPkcs8Base64Url);
    return {
        ...envelope,
        signature: {
            alg: 'ED25519',
            keyId,
            value: signature,
        },
    };
}
export async function verifyEnvelopeEd25519(envelope, publicKeySpkiBase64Url) {
    if (!envelope.signature)
        return false;
    if (envelope.signature.alg !== 'ED25519')
        return false;
    const unsigned = {
        header: envelope.header,
        body: envelope.body,
    };
    const payload = canonicalizeEnvelope(unsigned);
    return ed25519Verify(payload, envelope.signature.value, publicKeySpkiBase64Url);
}
export async function verifyEnvelopeSignature(envelope, material) {
    if (!envelope.signature)
        return false;
    if (envelope.signature.alg !== material.alg)
        return false;
    if (material.alg === 'HS256') {
        return verifyEnvelopeHmac(envelope, material.secret);
    }
    return verifyEnvelopeEd25519(envelope, material.publicKey);
}
export async function verifyCanonicalPayloadSignature(payload, signature, material) {
    if (!signature)
        return false;
    if (signature.alg !== material.alg)
        return false;
    if (material.alg === 'HS256') {
        return verifyCanonicalPayloadHmac(payload, signature.value, material.secret);
    }
    return verifyCanonicalPayloadEd25519(payload, signature.value, material.publicKey);
}
export function validateEnvelope(envelope, nowMs = Date.now()) {
    const diagnostics = [];
    const header = envelope.header ?? {};
    const body = envelope.body ?? {};
    const version = typeof header.version === 'string' ? header.version : '';
    const messageId = typeof header.messageId === 'string' ? header.messageId : '';
    const sender = typeof header.sender === 'string' ? header.sender : '';
    const nonce = typeof header.nonce === 'string' ? header.nonce : '';
    const timestampMs = typeof header.timestampMs === 'number' ? header.timestampMs : 0;
    const ttlMs = typeof header.ttlMs === 'number' ? header.ttlMs : 0;
    const content = typeof body.content === 'string' ? body.content : '';
    if (version !== '7h3/0.1') {
        diagnostics.push({ level: 'error', message: `Unsupported protocol version '${version}'` });
    }
    if (!messageId.trim()) {
        diagnostics.push({ level: 'error', message: 'Missing messageId' });
    }
    if (!sender.trim()) {
        diagnostics.push({ level: 'error', message: 'Missing sender identity' });
    }
    if (!nonce.trim()) {
        diagnostics.push({ level: 'error', message: 'Missing nonce — replay protection requires a unique nonce per message' });
    }
    if (ttlMs <= 0) {
        diagnostics.push({ level: 'error', message: 'ttlMs must be greater than zero' });
    }
    if (timestampMs + ttlMs < nowMs) {
        diagnostics.push({ level: 'error', message: 'Message TTL expired' });
    }
    if (!content.trim()) {
        diagnostics.push({ level: 'warning', message: 'Empty content payload' });
    }
    return diagnostics;
}
export function createEnvelope(input) {
    const nowMs = input.nowMs ?? Date.now();
    return {
        header: {
            version: '7h3/0.1',
            messageId: input.messageId ?? `msg-${nowMs}-${Math.random().toString(36).slice(2, 10)}`,
            timestampMs: nowMs,
            ttlMs: input.ttlMs ?? 60_000,
            sender: input.sender,
            recipient: input.recipient,
            nonce: input.nonce ?? Math.random().toString(36).slice(2, 12),
        },
        body: {
            intent: input.intent,
            content: input.content,
            capability: input.capability,
            correlationId: input.correlationId,
        },
    };
}
